### ISSUE TITLE
[iOS] “Authentication error Try again” red banner displayed when pressing the “Already a subscriber?” button from IAP view

### ISSUE THREAD

**Issue by data-sync-user** [mozilla-mobile/mozilla-vpn-client#7134]
**Important details:**

* not reproducing if signing in with a new Fx account -> “Problem confirming subscription” screen is displayed;

**Affected versions:**

* 2.16.0 (2.202306071809);

**Tested Platforms:**

* iOS 16.5;

**Prerequisites:**

* Have Mozilla VPN app installed and open;

* Have a Fx account with an expired IAP subscription (made on iOS);

**Steps to reproduce:**

1. Press the “Sign up” / “Already a subscriber?” button from the onboarding screen;
1. Sign in with the account from prerequisites;
1. Press the “Already a subscriber?” button from the IAP view;
1. Observe the behavior;

**Expected result:**

* “Problem confirming subscription / Sorry we are unable to connect your Firefox Account to a current subscription. Please try again or contact our support team for further assistance." error screen is displayed. The user remains logged in;

**Actual result:**

* The user is logged out and “Authentication error Try again” red banner is displayed;

**Notes:**

* See the attached video and logs;

┆Issue is synchronized with this [Jira Bug](https://mozilla-hub.atlassian.net/browse/VPN-4971)

**Comment by data-sync-user** [comment]
➤ Lesley Norton commented:

What happens when after re-authenticating?

**Comment by data-sync-user** [comment]
➤ Bianca Hidecuti commented:

Lesley Norton, after re-authenticating the user is redirected to the IAP view - where the subscriptions plans are displayed. The user is logged out only after pressing the “Already a subscriber?” button from the IAP view, while being logged in with a Fx account with an expired IAP subscription (the subscription is no longer active).

**Comment by data-sync-user** [comment]
➤ Matt Cleinman commented:

In `iosiaphandler.mm` :

        int errorNumber = errorValue.toInt();
        if (errorNumber == GUARDIAN_ERROR_RECEIPT_NOT_VALID) {
          REPORTNETWORKERROR(error, ErrorHandler::PropagateError, purchase->name());
          emit subscriptionFailed();
          emit ErrorHandler::instance()->subscriptionExpired();
          return;
        }I'm thinking either those two emits conflict somehow, or somewhere it fails out to the main screen and we don’t want it to.

Tracing subscriptionExpired, which is only picked up in one place:

  * In mozillavpn.cpp, it signals a request of ScreenSubscriptionExpiredError - I believe this is what we want to happen.

Tracing subscriptionFailed, which is picked up in two places:

  * In telemetry.cpp, where it records some Glean metrics.
  * In mozillavpn.cpp, where

**Comment by data-sync-user** [comment]
➤ Matt Cleinman commented:

(continuing my debug notes, feel free to ignore)

  * TaskGetLocation is setup in MozillaVPN::scheduleRefreshDataTasks. This is related to the new-ish suggested servers code, so it makes sense this was possibly a regression that was added somewhat recently.
  * scheduleRefreshDataTasks is called in a few spots, but in this case (from using breakpoints, we know it comes from MozillaVPN::initialize.
  * Looking into scheduleRefreshDataTasks, it also spawns the TaskAccount. I think - but am not sure - that we need TaskAccount to return first, as we need to get the bad subscription state before the location check happens.
  * Interesting… in scheduleRefreshDataTasks we schedule TaskGetLocation before we schedule the TaskGroup that includes TaskAccount. Moving that around and… it does not fix the issue.
  * While that builds, let’s dig into TaskGroup, because maybe that messes with scheduling in some way…
    * Yeah, that shouldn’t mess anything up. And in logs, looks like it’s working as expected.

  * OK, let’s gut check I'm in the right spot. I’m commenting out the TaskGetLocation work from scheduleRefreshDataTasks - if it works as expected, it means it is indeed this code. If not, it’s something else.
    * Result: Eeek, we still just fall back to the main authentication screen.
    * Theories: 
      * There is another place doing the same thing - and we should look at what calls setUserState to an unauthenticated user.
      * Alternatively, we’re just not doing the IAP check at the right spot. Let’s check how we confirm IAP subscriptions are active.

These last two theories are where I’ll pick up next week.

**Comment by data-sync-user** [comment]
➤ Matt Cleinman commented:

If TaskGetLocation is commented out, where is the failure coming from? Is this happening from another task, like TaskCaptivePortalLookup or similar?

Some time digging into the Guardian code to see why this might be happening on this call. More Guardian research into what can cause a 401. While Qt is interpreting the server response as something that throws a QNetworkReply::AuthenticationRequiredError, my guess is that this 401 is something that Guardian considers requireSubscriptionClient  - and it’s running quicker than we’d want it to in this flow.

In the logs - “state changed”. This only comes from IPAddressLookup. Which is a server call that requires an active subscription. And the IP address is triggered by a state change. Let’s see if this is the root cause.

Could also be TaskAddDevice and TaskRemoveDevice - those require an active subscription on the Guardian server - but seems less likely.

Commenting out IP finding calls to see what happens.

Not that - it’s TaskAccount - the first task of these - which is returning with a 401. I'm not quite sure why. Will chat w/ Jeremy tomorrow on this.

**Comment by data-sync-user** [comment]
➤ Matt Cleinman commented:

So we’re hitting this on TaskGetLocation, which hits the IPInfo endpoint, which needs an authenticated and active subscription user. But according to a comment in mozilla.cpp, this ordering is intentional - 

// The VPN needs to be off in order to determine the client's real location.
// And it also needs to complete before TaskServers in case this triggers an
// automatic server selection.
//
// TODO: This ordering requirement can be relaxed in the future once automatic
// server selection is implemented upon activation. See JIRA issue
// https://mozilla-hub.atlassian.net/browse/VPN-3726 for more information.  Can we just check for an expired subscription on this call, like we do in IOSIAPHandler::processCompletedTransactions? (Perhaps this should be handled at a lower level, like NetworkRequest::replyFinished()- but for now let’s not go that deep.)

After looking at Guardian code and talking to Jeremy, we should be looking for Guardian to throw a NoActiveSubscription, which we can identify on the client by the errno being 123. 

However, this isn’t catching the issue. Instead, it looks like the errno is 122. Why is that? 

Digging into Guardian’s code, this is a DeviceNotFound error, which is only thrown in util’s verifyJwtToken. Looking at that code, we check for a valid user, then check for a valid device. However, we also clean up devices of ended subscriptions, and so we throw an invalid device. It seems like we should check for an active sub before checking for a valid device, and throw a NoActiveSubscription (a 123) if there is no subscription. In other words: We’re throwing a valid, yet imprecise error (and an error which is less helpful to the client). I’m gut checking this w/ Jeremy to make sure this is something reasonable for Guardian to do in this spot, and won’t have any unforeseen consequences.

So, this is a two part fix:

  * Add this code to TaskGetLocation’s error handling in the client
  * On Guardian, check for a valid subscription in verifyJwtToken

Except that this doesn’t totally fix it… more soon.

**Comment by data-sync-user** [comment]
➤ Matt Cleinman commented:

So what is happening is that scheduleRefreshDataTasks kicks off five tasks. This is run right after starting the app. All five will return a 401, as verifyJwtToken will fail for each of them in the same way. We could put the same “if subscription expired, do this instead” code in the failure block of code for each task (or better yet, create a utility function to check this), but this seems like an inappropriately high level to put this in - it shouldn’t be up to each network request to check for this state. And this is brittle - any other network requests that are created or run in the app startup would cause this bug again.

Below these tasks is our NetworkRequest class. This class knows nothing about it’s requests, and shouldn’t be responsible for forking logic when a request errors because the subscription is expired. It’s too low level to put this check in.

We probably should rework this so we have a middle layer between these two that can handle situations like this. That’s a bigger lift than I was thinking with this one - and I have some blockers I need to focus at the moment.

(Also, this work requires the Guardian-side work described below.)

**Comment by data-sync-user** [comment]
➤ Santiago Andrigo commented:

Moving this to Kookaburra, since for some reason it didn’t migrate correctly there.

**Comment by data-sync-user** [comment]
➤ Matt Cleinman commented:

It’s highly possible this bug has been around for a while. My consistent repro steps (on staging):
1) Sign in with an account w/o a subscription.
2) Do a (sandbox) IAP on iOS
3) Cancel the sandbox subscription
4) Wait a few minutes to let the sub end
5) Cold launch of app, we end up on the blue screen w/ an auth error, rather than the nice “subscription expired” screen.

When the app launches, it runs five tasks as part of the boot:

  * TaskAccount
  * TaskServers
  * TaskCaptivePortalLookup
  * TaskGetSubscriptionDetails
  * TaskGetLocation

All 5 seem to return a 401 w/ errno 122 (which corresponds to Guardian’s DeviceNotFound). This makes sense - as the ended sub meant the device was scrubbed. It looks like that error is only thrown on Guardian in verifyJwtToken, which is checked for most calls.

I’m open to fixing it on client, but I’m not sure how we’re able to distinguish this error from a situation where the device was removed from an active account (perhaps because a user got the ‘too many devices’ screen on a sixth device). We certainly don’t want to show the “resubscribe” message to a user with a valid subscription - and without more info from Guardian, I’m not sure how we can distinguish these two situations on the client. 

Thus, this will probably need a fix on Guardian, and likely the client as well.

**Comment by data-sync-user** [comment]
➤ Bianca Hidecuti commented:

Still reproducing while using the 2.17.1 version, on iOS.

**Comment by data-sync-user** [comment]
➤ Santiago Andrigo commented:

Matt Cleinmanhttps://mozilla-hub.atlassian.net/browse/VPN-5644 ( https://mozilla-hub.atlassian.net/browse/VPN-5644|smart-link ) is an adjacent bug and likely one that could be fixed along with this one whenever we take it. 

Having said that, why is it marked as blocked? Blocked on what?

**Comment by data-sync-user** [comment]
➤ Santiago Andrigo commented:

Is there a ticket for this? Can you link it as ‘blocked by' so that it’s tracked?

**Comment by data-sync-user** [comment]
➤ Matt Cleinman commented:

Guardian-side work broken out  as https://mozilla-hub.atlassian.net/browse/VPN-5644 ( https://mozilla-hub.atlassian.net/browse/VPN-5644|smart-link ).
